lib/dice/src/cert.rs

```diff
@@ -135,9 +135,78 @@ pub trait CertBuilder {
     }
 }
 
-pub struct DeviceIdSelfCertBuilder([u8; deviceid_cert_tmpl::SIZE]);
+pub struct PersistIdSelfCertBuilder([u8; persistid_cert_tmpl::SIZE]);
 
-impl DeviceIdSelfCertBuilder {
+impl PersistIdSelfCertBuilder {
+    pub fn new(
+        cert_sn: &CertSerialNumber,
+        dname_sn: &SerialNumber,
+        public_key: &PublicKey,
+    ) -> Self {
+        Self(persistid_cert_tmpl::CERT_TMPL.clone())
+            .set_serial_number(cert_sn)
+            .set_issuer_sn(dname_sn)
+            .set_subject_sn(dname_sn)
+            .set_pub(public_key.as_bytes())
+    }
+
+    const SIGNDATA_RANGE: Range<usize> = persistid_cert_tmpl::SIGNDATA_RANGE;
+
+    pub fn sign(self, keypair: &Keypair) -> PersistIdSelfCert
+    where
+        Self: Sized,
+    {
+        let signdata = &self.0[Self::SIGNDATA_RANGE];
+        let sig = keypair.sign(signdata);
+        let tmp = self.set_sig(&sig.to_bytes());
+
+        PersistIdSelfCert(tmp.0)
+    }
+}
+
+impl CertBuilder for PersistIdSelfCertBuilder {
+    const SERIAL_NUMBER_RANGE: Range<usize> =
```

Comment:
This is probably my inexperience with Rust showing, or maybe I'm misunderstanding but wouldn't this break the functions implemented in the CertBuilder trait that use the constants? I've always viewed this trait as existing and using these constants so we don't have to implement the `set_*` functions in each of the concrete types.

Reply:
I don't believe this will cause an issue. The functions in the `CertBuilder` trait will just use the constants from the environment (i.e. the file). This is the same is as importing and using a library function in those trait methods.

The reason you would want to allow redefining constants per implementer is if you expected them to be different for each implementer. In this case, AFAICT, they are all the same and so there is no need for them to be parameterized.

Reply:
Oh I definitely misunderstood. Some of these constants are the same (like the serial number), some aren't. I went this way to be "consistent" which may be a synonym with "tedious" here :). I'm going to push the changes I've made per your comments here and think on this one a bit ... my gut reaction is that constants from the tmpls that should be consistent across all tmpls should probably come from somewhere else ...

Reply:
No problem. This turned out to be a fruitful discussion then.  It's up to you what to do about it. Adding the constants doesn't feel crazy hard to me. I was just pointing out if it's not needed, then we shouldn't do it. But I'm fine merging either way.